substrate/client/service/src/builder.rs

```diff
@@ -263,12 +265,80 @@ where
 			},
 		)?;
 
+		if let Some(warm_up_strategy) = config.warm_up_trie_cache {
+			let storage_root = client.usage_info().chain.best_hash;
+			let backend_clone = backend.clone();
+
+			if warm_up_strategy.is_blocking() {
+				// We use the blocking strategy for testing purposes.
+				// So better to error out if it fails.
+				warm_up_trie_cache(backend_clone, storage_root)?;
+			} else {
+				task_manager.spawn_handle().spawn_blocking(
+					"warm-up-trie-cache",
+					None,
+					async move {
+						if let Err(e) = warm_up_trie_cache(backend_clone, storage_root) {
+							error!("Failed to warm up trie cache: {e}");
+						}
+					},
+				);
+			}
+		}
+
 		client
 	};
 
 	Ok((client, backend, keystore_container, task_manager))
 }
 
+fn child_info(key: Vec<u8>) -> Option<ChildInfo> {
+	let prefixed_key = PrefixedStorageKey::new(key);
+	ChildType::from_prefixed_key(&prefixed_key).and_then(|(child_type, storage_key)| {
+		(child_type == ChildType::ParentKeyId).then(|| ChildInfo::new_default(storage_key))
+	})
+}
+
+fn warm_up_trie_cache<TBl: BlockT>(
+	backend: Arc<TFullBackend<TBl>>,
+	storage_root: TBl::Hash,
+) -> Result<(), Error> {
+	use sc_client_api::backend::Backend;
+	use sp_state_machine::Backend as StateBackend;
+	let state = || backend.state_at(storage_root, TrieCacheContext::Untrusted);
+	let keys = KeysIter::<_, TBl>::new(state()?, None, None)?.collect::<Vec<_>>();
```

Comment:
I don't think its a good idea to load all the keys into memory up front just to be able to calculate progress. I think it should be possible to just iterate directly without collecting since you don't need mutable access anywhere.

Reply:
Yeah, because now it's in background, we can skip the progress and just print the final time.

Reply:
Agreed. My point is just to get rid of the collect here.